app/helpers/api_helper.rb

```diff
@@ -0,0 +1,41 @@
+module ApiHelper
+  # Shape for a paginated response
+  def paginated(json, collection)
+    json.data do
+      yield
+    end
+    json.links do
+      json.pages! collection, url: request.original_url
+    end
+    json.meta do
+      json.count collection.total_count
+    end
+  end
+
+  # Shape for an object
+  def obj(json, object)
+    json.id object.hashid
+    json.type object.class.try(:api_type) || object.class.name.underscore
+
+    yield if block_given?
+
+    json.created_at object.created_at if object.respond_to?(:created_at)
+    json.links do
+      json.self api_url_for(object)
+    end
+  end
+
+  # API URL for an object. By default, the api_version of the generated URL is
+  # the same as the current request's version.
+  def api_url_for(object, **options)
+    polymorphic_url([:api, object], api_version: @request_version, **options)
+  end
+
+  # Permanent URL for an attached file or variant
+  def file_url_for(attachable, variant = nil)
```

Comment:
ActiveStorage provides link generators iirc

Reply:
### [`ActionView::Helpers.url_for`](https://apidock.com/rails/ActionView/Helpers/UrlHelper/url_for)
```ruby
url_for hackathon.logo
```
Doesn't provide the full URL. By default, it's `only_path: true`. However, I can't seem to properly pass in a `host` to get the full URL. Note that this implementation is actually different than `Rails.application.routes.url_helpers.url_for` from `ActionDispatch::Routing::RouteSet`

### [`ActionDispatch::Routing::RouteSet.url_for`](https://api.rubyonrails.org/classes/ActionDispatch/Routing/UrlFor.html)

```ruby
Rails.application.routes.url_helpers.url_for(hackathon.logo)
```

This is what we want! It's kinda of long tho

### `polymorphic_url`

```ruby
polymorphic_url(hackathon.logo)
```

This also works as expected! Full path and everything

---

So as a result, either `Rails.application.routes.url_helpers.url_for` or `polymorphic_url` works. Inside `file_url_for` I had provided a param to make variants cleaner.

```ruby
# before
file_url_for hackathon.logo.variant(:small)

# after
file_url_for hackathon.logo, :small
```